reframe/core/containers.py

```diff
@@ -0,0 +1,109 @@
+import abc
+
+import reframe.core.fields as fields
+import reframe.utility.typecheck as typ
+from reframe.core.exceptions import ContainerError
+
+
+class ContainerPlatform(abc.ABC):
+    """The abstract base class of any container platform.
+
+    Concrete container platforms inherit from this class and must override the
+    :func:`emit_prepare_cmds` and :func:`emit_launch_cmds` abstract functions.
+    """
+
+    registry = fields.TypedField('registry', str, type(None))
+    image = fields.TypedField('image', str, type(None))
+    requires_mpi = fields.TypedField('requires_mpi', bool)
+    commands = fields.TypedField('commands', typ.List[str])
+    mount_points = fields.TypedField('mount_points',
+                                     typ.List[typ.Tuple[str, str]])
+    workdir = fields.TypedField('workdir', str, type(None))
+
+    def __init__(self):
+        self.registry = None
+        self.image = None
+        self.requires_mpi = False
+        self.commands = []
+        self.mount_points  = []
+        self.workdir = None
+
+    @abc.abstractmethod
+    def emit_prepare_cmds(self):
+        """Returns commands that are necessary before running with this
+        container platform.
+
+        :raises: `ContainerError` in case of errors.
+
+        .. note:
+            This method is relevant only to developers of new container
+            platforms.
+        """
+
+    @abc.abstractmethod
+    def emit_launch_cmds(self):
+        """Returns the command for running with this container platform.
+
+        :raises: `ContainerError` in case of errors.
+
+        .. note:
+            This method is relevant only to developers of new container
+            platforms.
+        """
+        if self.registry:
+            self.image = '/'.join([self.registry, self.image])
+
+    @abc.abstractmethod
+    def validate(self):
+        """Validates this container platform.
+
+        :raises: `ContainerError` in case of errors.
+
+        .. note:
+            This method is relevant only to developers of new container
+            platforms.
+        """
+        if self.image is None:
+            raise ContainerError('no image specified')
+
+        if not self.commands:
+            raise ContainerError('no commands specified')
+
+
+class Docker(ContainerPlatform):
+    """An implementation of ContainerPlatform to run containers with Docker."""
+
+    def emit_prepare_cmds(self):
+        pass
+
+    def emit_launch_cmds(self):
+        super().emit_launch_cmds()
+        docker_opts = ['-v "%s":"%s"' % mp for mp in self.mount_points]
```

Comment:
You should use `--mount` as described here https://docs.docker.com/storage/bind-mounts/ since the stage dir is already created on the host machine.

Reply:
As discussed the `-v` option is ok for our case.